bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/cache/AudioStreamCacheWrapper.java

```diff
@@ -0,0 +1,189 @@
+            try {
+                byte[] bytesRead = ttsResult.read(offset, 1);
+                if (bytesRead.length == 0) {
+                    return -1;
+                } else {
+                    offset++;
+                    return bytesRead[0] & 0xff;
+                }
+            } catch (IOException e) {
+                logger.debug("Cannot read from tts cache. Will use the fallback TTS mechanism", e);
+            }
+        }
+
+        // beyond this point, we failed, so the fallback must be active :
+        enableFallback();
+        AudioStream fallbackDirectAudioStreamFinal = fallbackDirectAudioStream;
+        if (fallbackDirectAudioStreamFinal != null) {
+            return fallbackDirectAudioStreamFinal.read();
+        }
+
+        throw new IOException("Neither TTS cache nor TTS fallback method succeed");
+    }
+
+    @Override
+    public int read(byte @Nullable [] b, int off, int len) throws IOException {
+        if (fallbackDirectAudioStream == null) {
+            if (b == null) {
+                throw new IOException("Array to write is null");
+            }
+            Objects.checkFromIndexSize(off, len, b.length);
+
+            if (len == 0) {
+                return 0;
+            }
+
+            try {
+                byte[] bytesRead = ttsResult.read(offset, len);
+                offset += bytesRead.length;
+                if (bytesRead.length == 0) {
+                    return -1;
+                }
+                int i = 0;
+                for (; i < len && i < bytesRead.length; i++) {
+                    b[off + i] = bytesRead[i];
+                }
+                return i;
+            } catch (IOException e) {
+                logger.debug("Cannot read from tts cache. Will use the fallback TTS mechanism", e);
+            }
+        }
+
+        // beyond this point, we failed, so fallback must be active :
+        enableFallback();
+        AudioStream fallbackDirectAudioStreamFinal = fallbackDirectAudioStream;
+        if (fallbackDirectAudioStreamFinal != null) {
+            return fallbackDirectAudioStreamFinal.read(b, off, len);
+        }
+
+        throw new IOException("Neither TTS cache nor TTS fallback method succeed");
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        offset += n;
+        return n;
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            super.close();
+        } finally {
+            ttsResult.closeAudioStreamClient();
+        }
+    }
+
+    private void enableFallback() throws IOException {
+        if (fallbackDirectAudioStream == null) {
+            try {
+                this.fallbackDirectAudioStream = fallbackDirectSupplier.fallBackDirectResolution();
+                this.fallbackDirectAudioStream.skip(offset);
+            } catch (TTSException e) {
+                throw new IOException("Cannot read from TTS service", e);
+            }
+        }
+    }
+
```

Comment:
This is a backup call to the TTS service, and it should be used only if absolutely needed. So if we do it in the constructor, it negates the usefulness of the cache entirely ?

The purpose of the fallback mechanism is to deal with error, and above all the case of a suppression of the cache files.

Reply:
But you create the `fallBackAudioStream` only once and then re-use it, not every time you call `enableFallback`. So it's a one-time operation that does not affect the cache itself.

Reply:
I think the miscomprehension comes from the `read` method ?
I call `enableFallback()` in the `read` methods, which then set the `fallbackAudioStream`. But `enableFallback` is called ONLY if reading the file fails.
The code is not very clear, but `read()` should return data before coming to execute the `enableFallback()` method.

`enableFallback()`, under normal operation, should not be called, because it calls the `TTSService` (with a call to `fallbackDirectSupplier.fallBackDirectResolution()`), and the purpose of the cache is to avoid that.